pkg/pgmodel/ingestor/trace/trace_batcher.go

```diff
@@ -0,0 +1,283 @@
+		spanCount: 0,
+		reqStatus: make([]chan error, 0, maxBatchSize),
+		ctx:       context.Background(),
+		maxSize:   maxBatchSize, // this is not hard limit
+	}
+}
+
+func (tb *Batch) add(in insertTracesReq) {
+	_, addReqSpan := tracer.Default().Start(tb.ctx, "add-trace-req-to-batch",
+		trace.WithLinks(trace.Link{SpanContext: in.spanCtx}))
+	defer addReqSpan.End()
+	inSpans := in.payload.SpanCount()
+	if inSpans == 0 {
+		return
+	}
+	in.payload.ResourceSpans().MoveAndAppendTo(tb.traces.ResourceSpans())
+	tb.spanCount += inSpans
+	tb.reqStatus = append(tb.reqStatus, in.response)
+	tb.addCounter++
+}
+
+func (tb *Batch) isFull() bool {
+	return tb.spanCount >= tb.maxSize
+}
+
+func (tb *Batch) isEmpty() bool {
+	return tb.spanCount == 0
+}
+
+// Batcher batches trace requests and sends batches to batch writer.
+// This is done to achieve better ingest performance especially b/c
+// Jaeger collector sends traces one by one.
+type Batcher struct {
+	in              []chan insertTracesReq
+	stop            chan struct{}
+	batchWriter     *batchWriter
+	once            sync.Once
+	bufferedBatches chan Batch
+	wg              sync.WaitGroup
+	config          BatcherConfig
+}
+
+type BatcherConfig struct {
+	Batchers     int
+	Writers      int
+	MaxBatchSize int
+	BatchTimeout time.Duration
+}
+
+func NewBatcher(config BatcherConfig, writer Writer) *Batcher {
+	validateConfig(&config)
+	bufferedBatches := make(chan Batch, config.Writers*2) // we want some buffer to avoid writer waiting on batcher
+	inChs := make([]chan insertTracesReq, config.Batchers)
+	for i := range inChs {
+		inChs[i] = make(chan insertTracesReq, config.MaxBatchSize*3) // buffer for incoming requests, especially important for async acks
+	}
+	return &Batcher{
+		batchWriter:     newBatchWriter(config.Writers, writer, bufferedBatches),
+		in:              inChs,
+		stop:            make(chan struct{}),
+		bufferedBatches: bufferedBatches,
+		config:          config,
+	}
+}
+
+func (b *Batcher) send(req insertTracesReq, batcherIdx int) {
+	b.in[batcherIdx] <- req
+}
+
+func validateConfig(config *BatcherConfig) {
+	if config.Batchers == 0 {
+		config.Batchers = DefaultBatchWorkers
+	}
+	if config.MaxBatchSize == 0 {
+		config.MaxBatchSize = DefaultBatchSize
+	}
+	if config.BatchTimeout == 0 {
+		config.BatchTimeout = DefaultBatchTimeout
+	}
+	if config.Batchers < 1 || config.Writers < 1 {
+		panic("number of batchers and writeres must be greater then zero")
+	}
+}
+
+func (b *Batcher) Run() {
+	for i := 0; i < b.config.Batchers; i++ {
+		b.wg.Add(1)
+		go func(idx int) {
+			defer b.wg.Done()
+			b.batch(idx)
+		}(i)
+	}
+	b.batchWriter.run()
+}
+
+func (b *Batcher) batch(batchIdx int) {
+	_, batcherSpan := tracer.Default().Start(context.Background(), "create-batches")
+	ticker := time.NewTicker(b.config.BatchTimeout)
+	defer ticker.Stop()
+	batch := NewBatch(b.config.MaxBatchSize)
+	batcherSpan.AddEvent("New Batch")
+	flushBatch := func(batch *Batch) *Batch {
```

Comment:
During a flush, I'd consider stopping the ticker or resetting it to a really long period. It shouldn't really fire until reset again on first insert.